File: apps/ios/TaakHelden/App/AppConfiguration.swift
import Foundation

/// Build-time / runtime API base URL resolution for App Store-safe builds.
///
/// Order: process environment override → Info.plist (`TAAKHELDEN_API_BASE_URL`) →
/// production Workers URL. Never falls back to localhost (that broke Review builds).
enum AppConfiguration {
    /// Public staging/production API used when Info.plist is missing the key.
    static let productionAPIBaseURLString = "https://taakhelden-api.oostelaar.workers.dev/v1"

    static func apiBaseURL(
        processEnvironment: [String: String] = ProcessInfo.processInfo.environment,
        infoDictionary: [String: Any]? = Bundle.main.infoDictionary
    ) -> URL {
        if let raw = processEnvironment["TAAKHELDEN_API_BASE_URL"]?.trimmingCharacters(in: .whitespacesAndNewlines),
           !raw.isEmpty,
           let url = URL(string: raw)
        {
            return url
        }
        if let raw = infoDictionary?["TAAKHELDEN_API_BASE_URL"] as? String {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            // Unexpanded `$(TAAKHELDEN_API_BASE_URL)` must not win over the safety net.
            if !trimmed.isEmpty, !trimmed.hasPrefix("$("), let url = URL(string: trimmed) {
                return url
            }
        }
        return URL(string: productionAPIBaseURLString)!
    }
}

File: apps/android/core/src/main/kotlin/nl/taakhelden/core/config/AppConfiguration.kt
package nl.taakhelden.core.config

/**
 * Resolves the API base URL for release-safe builds.
 *
 * Order: explicit override (build config / environment) → production Workers URL. It
 * never falls back to localhost — on iOS that once shipped in a Review build and broke
 * it, so the safety net here is always the public API.
 */
public object AppConfiguration {
    public const val PRODUCTION_API_BASE_URL: String =
        "https://taakhelden-api.oostelaar.workers.dev/v1"

    public const val OVERRIDE_ENV_KEY: String = "TAAKHELDEN_API_BASE_URL"

    public fun apiBaseUrl(
        buildConfigValue: String? = null,
        environment: Map<String, String> = System.getenv(),
    ): String {
        buildConfigValue?.trim()?.let { candidate ->
            // An unexpanded Gradle placeholder must not win over the safety net.
            if (candidate.isNotEmpty() && !candidate.startsWith("\$")) return candidate
        }
        environment[OVERRIDE_ENV_KEY]?.trim()?.let { candidate ->
            if (candidate.isNotEmpty()) return candidate
        }
        return PRODUCTION_API_BASE_URL
    }
}
